csi/node.go

```diff
@@ -30,3 +52,128 @@ func (s *OsdCsiServer) GetNodeID(ctx context.Context, req *csi.GetNodeIDRequest)
 
 	return result, nil
 }
+
+// NodePublishVolume is a CSI API call which mounts the volume on the specified
+// target path on the node.
+//
+// TODO: Support READ ONLY Mounts
+//
+func (s *OsdCsiServer) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+
+	dlog.Debugf("NodePublishVolume req[%#v]", req)
+
+	// Check arguments
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	}
+	if len(req.GetTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path must be provided")
+	}
+	if req.GetVolumeCapability() == nil || req.GetVolumeCapability().GetAccessMode() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume access mode must be provided")
+	}
+
+	// Get volume information
+	v, err := util.VolumeFromName(s.driver, req.GetVolumeId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Volume id %s not found: %s",
+			req.GetVolumeId(),
+			err.Error())
+	}
+
+	// Gather volume attributes
+	spec, _, _, err := s.specHandler.SpecFromOpts(req.GetVolumeAttributes())
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"Invalid volume attributes: %#v",
+			req.GetVolumeAttributes())
+	}
+
+	// This seems weird as a way to change opts to map[string]string
+	opts := make(map[string]string)
+	if len(spec.GetPassphrase()) != 0 {
+		opts[options.OptionsSecret] = spec.GetPassphrase()
+	}
+
+	// Verify target location is an existing directory
+	// See: https://github.com/container-storage-interface/spec/issues/60
+	err = verifyTargetLocation(req.GetTargetPath())
+	if err != nil {
+		return nil, status.Error(codes.NotFound, err.Error())
```

Comment:
Yes, correct. that is what is done here. The call `status.Error()` encapsulates a grpc error and a message as a golang error. So here we only use err.Error(), instead of nesting it in another formatted string.

Reply:
that is what I am saying, the verify method will return a grpc error and when you do err.Error() here it will already have a status code. Here we are adding one more layer.
This is what err.Error() will have:
`fmt.Sprintf("rpc error: code = %s desc = %s", codes.Code(p.GetCode()), p.GetMessage())`

Reply:
Oh! Gotcha. In fact I think that verifyTargetLocation should _not_ send back grpc error. Now that I think about it, it should send back a normal err so that other non-grpc calls can use it later. I'll make this change.